### PR TITLE
MueLu: Amgx rebuild

### PR DESCRIPTION
**Pull request by shkevin**
This issue was due to the AMGX_initialize being done for every
AMGXOperator object. AMGX requires a single instance of their object.
The fix was removing the initialization and placing the init into the
Driver.cpp file.

@trilinos/<MueLu>

## Description
Extracted AMGX_initialize to fix MueLu+AMGX error.

## Motivation and Context
Couldn't specify --rebuild option for AMGX. Using this option will allow me to better benchmark MueLu vs AMGX.

## How Has This Been Tested?
This was tested by reconfiguring, rebuilding, and then running with various numRebuilds. The code would segfault with 50 rebuilds, but this was a different issue than the one solved.

### CODE REVIEW

**Comment by mhoemmen** [comment]
@trilinos/muelu

File: packages/muelu/adapters/amgx/MueLu_AMGX_Setup.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+//
+// @HEADER
+
+#include "MueLu_AMGX_Setup.hpp"
+
+namespace MueLu {
+
+  void MueLu_AMGX_initialize()
+  {
+    printf("************************AMGX_initialize********************************\n");
```
Comment (csiefer2) [inline packages/muelu/adapters/amgx/MueLu_AMGX_Setup.cpp]: Please remove printf

File: packages/muelu/adapters/amgx/MueLu_AMGX_Setup.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  void MueLu_AMGX_initialize_plugins()
+  {    
+    printf("********************AMGX_initialize_plugins****************************\n");
+    AMGX_SAFE_CALL(AMGX_initialize_plugins());
+  }
+
+  void MueLu_AMGX_finalize()
+  {    printf("**********************AMGX_finalize*********************************\n");
```
Comment (csiefer2) [inline packages/muelu/adapters/amgx/MueLu_AMGX_Setup.cpp]: Please remove printf

File: packages/muelu/adapters/amgx/MueLu_AMGX_Setup.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    AMGX_SAFE_CALL(AMGX_finalize());
+  }
+
+  void MueLu_AMGX_finalize_plugins()
+  {    
+    printf("********************AMGX_finalize_plugins******************************\n");
```
Comment (csiefer2) [inline packages/muelu/adapters/amgx/MueLu_AMGX_Setup.cpp]: Please remove printf

File: packages/muelu/adapters/amgx/MueLu_AMGX_Setup.hpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+//
+// Questions? Contact
+//                    Jonathan Hu       (jhu@sandia.gov)
+//                    Andrey Prokopenko (aprokop@sandia.gov)
+//                    Ray Tuminaro      (rstumin@sandia.gov)
+//
+// ***********************************************************************
+//
+// @HEADER
+#ifndef MUELU_AMGX_SETUP_DECL_HPP
```
Comment (csiefer2) [inline packages/muelu/adapters/amgx/MueLu_AMGX_Setup.hpp]: Please remove "DECL" from the ifndef / define so it matches the file name.

File: packages/muelu/adapters/amgx/MueLu_AMGX_Setup.hpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+#define MUELU_AMGX_SETUP_DECL_HPP
+
```
Comment (csiefer2) [inline packages/muelu/adapters/amgx/MueLu_AMGX_Setup.hpp]: This should be guarded with
#ifdef HAVE_MUELU_AMGX

File: packages/muelu/test/scaling/Driver.cpp [hunk 6]
```diff
@@ -118,7 +118,7 @@
 #include "Xpetra_EpetraMultiVector.hpp"
 #endif
 
-
+#include <MueLu_AMGX_Setup.cpp>
```
Comment (csiefer2) [inline packages/muelu/test/scaling/Driver.cpp]: You should be including the header, not the cpp

File: packages/muelu/test/scaling/Driver.cpp [hunk 7]
```diff
@@ -535,6 +541,12 @@ int main_(Teuchos::CommandLineProcessor &clp, Xpetra::UnderlyingLib& lib, int ar
 
   }//end reruns
 
+#ifdef HAVE_MUELU_AMGX
+// Finalize AMGX
+//MueLu::MueLu_AMGX_finalize();
+//MueLu::MueLu_AMGX_finalize_plugins();
+#endif
```
Comment (csiefer2) [inline packages/muelu/test/scaling/Driver.cpp]: Is there a reason you're not finalizing?

File: srcTarball.txt [hunk 8]
```diff
@@ -0,0 +1,11 @@
+AMGX/CMakeLists.txt
```
Comment (csiefer2) [inline srcTarball.txt]: These changes should not be here.

File: packages/muelu/test/scaling/Driver.cpp (hunk 7 above)
Comment (shkevin) [reply packages/muelu/test/scaling/Driver.cpp]: Finalizing led to memory leak errors. I assume this is why the person who wrote the original MueLu_AMGXOperator_decl.hpp didn't include finalize either.

**Comment by jhux2** [comment]
___
*[packages/muelu/adapters/amgx/MueLu_AMGXOperator_decl.hpp, line 186 at r2](https://reviewable.io/reviews/trilinos/trilinos/5339#-LgiWGR2145Gx56BvP0Z:-LgiWGR36NwNeI38SAOp:bnyrr8j) ([raw file](https://github.com/trilinos/trilinos/blob/0059dda234487eb6b3972d61bcdb7ae5caff30af/packages/muelu/adapters/amgx/MueLu_AMGXOperator_decl.hpp#L186)):*
> ```cpp
>       // Initialize
>       //AMGX_SAFE_CALL(AMGX_initialize());
>       //AMGX_SAFE_CALL(AMGX_initialize_plugins());
> ```

Do you want to keep these commented out as a reminder?  Otherwise, git remembers all, so they can be deleted.

**Comment by jhux2** [comment]
___
*[packages/muelu/test/scaling/Driver.cpp, line 548 at r2](https://reviewable.io/reviews/trilinos/trilinos/5339#-LgiVruF0yOwvzrExAL2-r2-548:-LgiWan70axZSziwuEDh:b919i4l) ([raw file](https://github.com/trilinos/trilinos/blob/0059dda234487eb6b3972d61bcdb7ae5caff30af/packages/muelu/test/scaling/Driver.cpp#L548)):*
<details><summary><i>Previously, shkevin wrote…</i></summary><blockquote>

Finalizing led to memory leak errors. I assume this is why the person who wrote the original MueLu_AMGXOperator_decl.hpp didn't include finalize either. 
</blockquote></details>

Leaks in MueLu or AMGX?

**Pull request merged**